src/backend.md

```diff
@@ -486,6 +486,56 @@ One of the key principles in effective transaction management is the avoidance o
    }
    ```
 
+1. Spray job should include all preconditions and call an action job for an entity when this entity is eligible for this action right now. The action job should be a thin wrapper. Otherwise, the spray job will invoke the action everytime and then our the queue is filled with undoable and duplicate tasks. The infrastructure wastes additionall resources processing useless work.
+
+   ```typescript
+   // bad
+
+   async function sprayXJob() {
+     const items = await findAll({
+       // many preconditions
+     });
+     await callActionJobForItems(items);
+   }
+
+   async function actionXJob(id: number) {
+     const item = await findByPk(id);
+     if (!item) {
+       return;
+     }
+     if (!isNow(item)) {
+       return;
+     }
+     if (!isX(item)) {
+       return;
+     }
+     if (!isY(item)) {
+       return;
+     }
+     doMainAction(item);
+   }
+
+   // good
+
+   async function sprayXJob() {
+     const items = await findAll({
+       // many preconditions include
+       // isX,
+       // isY,
+       // isNow,
+     });
+     await callActionJobForItems(items);
+   }
+
+   async function actionXJob(id: number) {
+     const item = await findByPk(id);
+     if (!item) {
+       return;
+     }
+     doMainAction(item);
```

Comment:
What do you think of duplicating all the preconditions in spray and in action jobs? i use to keep this preconditions as similar as possible. Mostly to prevent unexpected behaviour, if somebody, for unknown reasons, dicided to use my `actionJob` somewhere else, outside of `sprayJob`, by passing all the preconditions i have it sprayJob.

Plus, all the conditions that depends on time, they are relative (isNow). the time when sprayJob & actions execute such conditions might be different. in one moment it can be valid, and next moment - invalid, so its better to check these conditions all the time

Reply:
>What do you think of duplicating all the preconditions in spray and action jobs?  I used to keep these preconditions as similar as possible.

I used it too and have
- double tests because I must verify preconditions in spray and action jobs. We might de-synchronise the main preconditions without tests.
- we can reuse spray logic only as SQL(sequlize) where options. So we should make a SQL. The cron job is okay, but I'm not sure about the SNS job.

My main idea is to sync preconditions by spray and action jobs. I see that now we talk about different things. I changed the text.
now, does it make sense?

We can discuss the issue of action preconditions in the following PR, since we have cron and SNS action jobs.